✨ Source Harvest: Migrate to Low Code

pnilan opened the pull request:
## What
* Migrates Harvest to Low Code per How To Migrate A Python Connector To Low Code guide
* Unpins CDK, setting airbyte-cdk version `^0`
* Adds `config_migrations.py` to add new required `auth_type` property to config
* Adds unit tests for `config_migrations.py`
* Updates `test_source.py`, `unit_test.py`, and `test_streams.py` tests. Adds `freezegun` dev dependency.
* Bumps major version number, updates metadata, changelog, and harvest-migrations.md accordingly
* Closes https://github.com/airbytehq/airbyte-internal-issues/issues/6620

## 🚨 Breaking changes
* Incremental substreams now use per-partition state which is a breaking change affecting the following streams:
    - `estimate_messages`
    - `invoice_messages`
    - `invoice_payments`
    - `project_assignments`
 * Updates report stream slices to use 365-day slice durations. This affects following streams:
    - `expenses_clients`
    - `expenses_categories`
    - `expenses_projects`
    - `expenses_team`
    - `time_clients`
    - `time_projects`
    - `time_tasks`
    - `time_team`
    - `uninvoiced`

artem1205 commented on airbyte-integrations/connectors/source-harvest/source_harvest/spec.yaml:
```diff
@@ -0,0 +1,54 @@
+  required:
+    - account_id
+    - start_date
+    - auth_type
```
spec file does not align with the previous version.
credentials should be a property of type `object` with a required list depending on its type.

pnilan replied:
Using the connector builder changed the format of the spec. To keep it compatible with the connector builder for future updates, I felt like it is preferred to force the change now.

pnilan replied:
Reverted config to old format and added a `config_migration.py`.

artem1205 commented on airbyte-integrations/connectors/source-harvest/source_harvest/availability_strategy.py:
```diff
@@ -1,26 +0,0 @@
-        self, stream: Stream, logger: logging.Logger, source: Source, error: HTTPError
-    ) -> Dict[int, str]:
-        reasons_for_codes: Dict[int, str] = {
-            requests.codes.UNAUTHORIZED: "Please ensure your credentials are valid.",
-            requests.codes.FORBIDDEN: "This is most likely due to insufficient permissions on the credentials in use.",
-            requests.codes.NOT_FOUND: "Please ensure that your account ID is properly set. If it is the case and you are still seeing this error, please contact Airbyte support.",
```
I think we should migrate this block to `error_handler`

maxi297 replied:
100% agree as else, we might get paged for things that are config_errors

pnilan replied:
Good call. Will update!

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        datetime_format: "%Y-%m-%dT%H:%M:%SZ"
+      step: P365D
+      cursor_granularity: P1D
+  - type: DeclarativeStream
```
Differences observed:
* The shape of the state is different as this is a "per partition" state. This means that this will induce a breaking change and this needs to be documented

pnilan replied:
Documented.

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+      cursor_field: updated_at
+      cursor_datetime_formats:
+        - "%Y-%m-%dT%H:%M:%SZ"
+      datetime_format: "%Y-%m-%dT%H:%M:%SZ"
```
This comment applies to all the `incremental_sync` that were previously implemented using `IncrementalHarvestStream`:
* Previous datetime format used was `%Y-%m-%dT%H:%M:%S%Z` instead of `%Y-%m-%dT%H:%M:%SZ` but I think it was an error so I'm comfortable with going with this change
* Without https://github.com/airbytehq/airbyte-internal-issues/issues/6314, the stream changes the value we take to update the state

pnilan replied:
Agreed that the previous datetime format was likely an error.

Looks like the state based on the most recent record has been implemented, now, correct?

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        type: RequestOption
+        field_name: updated_since
+        inject_into: request_parameter
+  - type: DeclarativeStream
```
Differences observed:
* The shape of the state is different as this is a "per partition" state. This means that this will induce a breaking change and this needs to be documented

girarda replied:
this stream doesn't support incremental, right? What's the breaking change?

pnilan replied:
Agreed @girarda

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        type: RequestOption
+        field_name: updated_since
+        inject_into: request_parameter
+  - type: DeclarativeStream
```
Differences observed:
* The shape of the state is different as this is a "per partition" state. This means that this will induce a breaking change and this needs to be documented

pnilan replied:
Documented.

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        type: RequestOption
+        field_name: updated_since
+        inject_into: request_parameter
+  - type: DeclarativeStream
```
Differences observed: same as `invoice_payments`

pnilan replied:
Documented.

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        datetime_format: "%Y-%m-%dT%H:%M:%SZ"
+      step: P365D
+      cursor_granularity: P1D
+  - type: DeclarativeStream
```
Differences observed:
* The previous implementation would use different boundaries `{"from": "20210101", "to": "20220101"}` instead of `{"from": "20210101", "to": "20211231"}`. As [Harvest query params include the boundaries](https://help.getharvest.com/api-v2/reports-api/reports/expense-reports/#clients-report), it seems fine but don't hesitate to do more testing around that

pnilan replied:
Fair point, I'll add testing to verify my assumptions.

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        datetime_format: "%Y-%m-%dT%H:%M:%SZ"
+      step: P365D
+      cursor_granularity: P1D
+  - type: DeclarativeStream
```
Differences observed: same as `expenses_clients`

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        type: RequestOption
+        field_name: updated_since
+        inject_into: request_parameter
+  - type: DeclarativeStream
```
Differences observed: same as `expenses_clients`

pnilan replied:
Documented.

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        datetime_format: "%Y-%m-%dT%H:%M:%SZ"
+      step: P365D
+      cursor_granularity: P1D
+  - type: DeclarativeStream
```
Differences observed: same as `expenses_clients`

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        datetime_format: "%Y-%m-%dT%H:%M:%SZ"
+      step: P365D
+      cursor_granularity: P1D
+  - type: DeclarativeStream
```
Differences observed: same as `expenses_clients`

pnilan replied:
Documented.

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+          cursor_value: '{{ response.get("links", {}).get("next", {}) }}'
+          stop_condition: '{{ not response.get("links", {}).get("next", {}) }}'
+      partition_router: []
+  - type: DeclarativeStream
```
Differences observed: same as `expenses_clients`

pnilan replied:
Documented.

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        datetime_format: "%Y-%m-%dT%H:%M:%SZ"
+      step: P365D
+      cursor_granularity: P1D
+  - type: DeclarativeStream
```
Differences observed: same as `expenses_clients`

pnilan replied:
Documented.

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        datetime_format: "%Y-%m-%dT%H:%M:%SZ"
+      step: P365D
+      cursor_granularity: P1D
+  - type: DeclarativeStream
```
Differences observed: same as `expenses_clients`

pnilan replied:
Documented.

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        datetime_format: "%Y-%m-%dT%H:%M:%SZ"
+      step: P365D
+      cursor_granularity: P1D
+  - type: DeclarativeStream
```
Differences observed: same as `expenses_clients`

pnilan replied:
Document.

maxi297 commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+              refresh_request_body: {}
+              token_refresh_endpoint: https://id.getharvest.com/api/v2/oauth2/token
+        error_handler:
+          type: CompositeErrorHandler
```
Why is this composite if there is only one? Is this a Connector Builder thing?

maxi297 commented on airbyte-integrations/connectors/source-harvest/unit_tests/unit_test.py:
```diff
@@ -1,51 +0,0 @@
-#
```
I feel like we should keep those tests. What was the reason for removing them? Maybe `test_next_page_token` needs more work and therefore I'm whiling to drop it but the rest should be pretty straighforward

pnilan replied:
@maxi297 

I guess I'm a bit confused why these tests would be necessary when this functionality is effectively already tested by the low code CDK unit tests as well as the CATs -- `invoices`, `invoice_payments`, and `expenses_clients` all return records for the sandbox account. Let me know your thoughts.

pnilan replied:
Added the tests back and updated to function with the new low-code implementation.

maxi297 reviewed (changes requested):
Can we make sure to document the breaking change? Some of my comments are just for you to acknowledge that there is a change in behavior and that you were expecting them

pnilan commented on airbyte-integrations/connectors/source-harvest/source_harvest/spec.yaml:
```diff
@@ -0,0 +1,54 @@
+  type: object
+  required:
+    - account_id
+    - start_date
```
Config Builder doesn't allow for custom start_date names, was adapting connector to builder limitations.

tolik0 replied:
@pnilan Please add configuration migration from the previous spec format. This will allow users to avoid re-authentication and re-entering all spec values. Here is [an example](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-google-ads/source_google_ads/config_migrations.py#L30). Additionally, please don't delete previous parameters to maintain backward compatibility.

pnilan replied:
@tolik0 Thanks for this. Added `config_migrations.py` and tests.

pnilan commented on airbyte-integrations/connectors/source-harvest/source_harvest/manifest.yaml:
```diff
@@ -0,0 +1,2832 @@
+        type: RequestOption
+        field_name: updated_since
+        inject_into: request_parameter
+  - type: DeclarativeStream
```
@maxi297 Thanks for the explanation. I noticed there were more records being returned with the low-code implementation but couldn't identify why that was happening.

How did you manage to discover that it's due to state being updated prematurely?

maxi297 replied:
Unfortunately, I don't have a good process for that. I got a hunch and tested it manually...

pnilan commented on airbyte-integrations/connectors/source-harvest/metadata.yaml:
```diff
@@ -26,9 +26,14 @@ data:
+  releases:
+    breakingChanges:
+      1.0.0:
+        message: "ThThis update requires a config update. Please update your config and reset your data (to match the new format). This version has added an additional field, requiring you to select either 'API Token' authentication or 'OAuth' authentication. Additionally, the replication_end_date option has been removed."
```
Fixed -- updated to reflect per-partition state breaking changes.

alafanechere reviewed (commented):
### Disclaimer
I did not look at the code at all and I'm experimenting with our regression tool. My goal is to assess if this tool is good enough to provide an holistic comparison of the connector behavior between versions.

### Conclusion
It looks very good. I did not spot any missing records.

### Stream coverage

I tested the following streams:

* billable_rates
* invoices
* time_entries
* time_tasks
* tasks
* project_assignments
* time_projects
* users
* invoice_item_categories
* cost_rates
* clients
* invoice_messages
* expense_categories
* company
* user_assignments
* projects
* time_team
* task_assignments
* roles
* time_clients
* uninvoiced
* project_budget

I've yet to test:
- contacts
- invoice_payments
- estimates
- estimate_messages
- estimate_item_categories
- expenses
- expenses_clients
- expenses_projects
- expenses_categories
- expenses_team
### Extra records on two streams
I spotted streams with more records in the new version:
* project assignment
* invoice message

### Slicing logic changing the record content
I think the slicing logic impacts the record content (the from and to field). Which would explain why on the following streams I have the exact same number of added and removed records, which is not altering the global record counts:
* time_clients
* time_project
* time_team
* uninvoiced

For instance I see that the current version calls:

`https://api.harvestapp.com/v2/reports/uninvoiced?from=20200101&per_page=50&to=20201231`
 While the new one calls
 
`https://api.harvestapp.com/v2/reports/uninvoiced?from=20200101&per_page=50&to=20201230`

**Do you think you could adapt the manifest to make sure the slicing logic matches the one of the current version? Making them match means there will be less drift in terms of record content when users will perform a reset of their sync to get the new state format.**

### Next steps
I'll test the missing streams and give you a final 👍 or 👎

pnilan commented:
@alafanechere 

The shift in slicing logic was intentional.

The report endpoints allow a [max 1-year (365 day) duration](https://help.getharvest.com/api-v2/reports-api/reports/expense-reports/) with inclusive boundaries. So valid slices would be [01/01/2021, 12/31/2021], [01/01/2022, 12/31/2021], etc.

Previously the connector was slicing from 01/01/2023 to 01/01/2024, or 366 day duration and then using the “from” date as the new “to” date, for example: [01/01/2021, 01/01/2022], [01/01/2022, 01/01/2023], etc. (In your example above, 01/01/2020 to 12/31/2020 is 366 days because 2020 was a leap year). Unclear if the API actually allows for 366-day durations for leap years but it does explicitly say 365-day max duration. 

```python
   while start_date < end_date:
            # Max size of date chunks is 365 days
            # Docs: https://help.getharvest.com/api-v2/reports-api/reports/time-reports/
            end_date_slice = end_date if start_date >= end_date.subtract(days=365) else start_date.add(days=365)
            date_slice = {"from": start_date.strftime(self.date_param_template), "to": end_date_slice.strftime(self.date_param_template)}

            start_date = end_date_slice

            yield date_slice
```
Python-implemented slicing logic.

Updating the low-code implementation will require me to [override the low-code slicing logic](https://github.com/airbytehq/airbyte/blob/aca8679a0e74a64e5b66cd205225bcc5d32487db/airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py#L167).

Given the above I think we should stick with the 365-day slice duration, but if you think it's more valuable to reflect the original implementation more closely, I'm happy to update. Let me know your thoughts.

katmarkham commented on airbyte-integrations/connectors/source-harvest/metadata.yaml:
```diff
@@ -26,9 +26,14 @@ data:
+  releases:
+    breakingChanges:
+      1.0.0:
+        message: "This update requires a reset for the following streams to due an update in the format of state: `expenses_clients`, `expenses_categories`, `expenses_projects`, `expenses_team`, `time_clients`, `time_projects`, `time_tasks`, `time_team`, `uninvoiced`, `estimate_messages`, `invoice_payments`, `project_assignments`."
```
Please start with "Several changes have been made to the Harvest connector."

pnilan replied:
Updated

katmarkham commented on docs/integrations/sources/harvest-migrations.md:
```diff
@@ -0,0 +1,30 @@
+- `invoice_payments`
+- `project_assignments`
+
+We're continuously striving to enhance the quality and reliability of our connectors at Airbyte.  As part of our commitment to delivering exceptional service, we are transitioning Source Harvest from the Python Connector Development Kit (CDK) to our a new low-code framework improving maintainability and reliability of the connector. However, due to differences between the Python and low-code CDKs ...
```
there is a typo::

We're continuously striving to enhance the quality and reliability of our connectors at Airbyte.  As part of our commitment to delivering exceptional service, we are transitioning Source Harvest from the Python Connector Development Kit (CDK) to our new low-code framework improving maintainability and reliability of the connector. However, due to differences between the Python and low-code CDKs, this migration constitutes a breaking change.

pnilan replied:
Updated.

alafanechere commented:
> @alafanechere
> 
> The shift in slicing logic was intentional.
> 
> The report endpoints allow a [max 1-year (365 day) duration](https://help.getharvest.com/api-v2/reports-api/reports/expense-reports/) with inclusive boundaries. So valid slices would be [01/01/2021, 12/31/2021], [01/01/2022, 12/31/2021], etc.
> 
> Previously the connector was slicing from 01/01/2023 to 01/01/2024, or 366 day duration and then using the “from” date as the new “to” date, for example: [01/01/2021, 01/01/2022], [01/01/2022, 01/01/2023], etc. (In your example above, 01/01/2020 to 12/31/2020 is 366 days because 2020 was a leap year). Unclear if the API actually allows for 366-day durations for leap years but it does explicitly say 365-day max duration.
> 
> ```python
>    while start_date < end_date:
>             # Max size of date chunks is 365 days
>             # Docs: https://help.getharvest.com/api-v2/reports-api/reports/time-reports/
>             end_date_slice = end_date if start_date >= end_date.subtract(days=365) else start_date.add(days=365)
>             date_slice = {"from": start_date.strftime(self.date_param_template), "to": end_date_slice.strftime(self.date_param_template)}
> 
>             start_date = end_date_slice
> 
>             yield date_slice
> ```
> 
> Originally implemented slicing logic. Uses previous end date as new start date.
> 
> Updating the low-code implementation will require me to [override the low-code slicing logic](https://github.com/airbytehq/airbyte/blob/aca8679a0e74a64e5b66cd205225bcc5d32487db/airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py#L167).
> 
> Given the above I think we should stick with the 365-day slice duration, but if you think it's more valuable to reflect the original implementation more closely, I'm happy to update. Let me know your thoughts.

I appreciate the detailed explanation. I'm in as long as we are aware of the reason of this behavioral change and you think it's more correct 👍 .
@pnilan do you think the `to` and `from` fields which are passed to records according to the requests parameters are of any use for customers? I'm wondering if we should remove them so that future changes to the slicing logic does not impact the destination.
I am asking now because it'd be a breaking change and might be interesting to bundle two breaking changes in the same release to avoid disturbing users multiple times.

alafanechere reviewed (approved):
Approving! From a behavioral standpoint our regression test tool did not detect any alarming anomaly, quite the opposite as I think you fixed missing records on the following streams:
* invoice_payments
* project_assignments
* invoice_messages
(you might want to add that to the release notes).

I'll let @katmarkham validate the breaking change messaging is correct.

pnilan commented:
@alafanechere 

I do think they are probably used in the destination as otherwise the report records do not provide any cursor values, so it could be used to track changes in reporting values over time.

See [Expense Reports documentation](https://help.getharvest.com/api-v2/reports-api/reports/expense-reports/#clients-report).

maxi297 reviewed (approved):
I haven't rechecked but I'll approve based on @alafanechere's approval to unblock the merge

pnilan commented:
@artem1205 

Updated and confirmed the new per-partition state causes a breaking change.

Your comment uncovered a potential issue with `test_state_with_abnormally_large_values` -- I was using the old state format but the test was passing, I even changed the date to be "reasonable" (non-abnormal?) and it would still pass. I could only get the test to fail once I updated to the correct new per-partition format and set the date to the a "reasonable" date. So now the format is updated for the relevant streams and I have confirmed that the test is functioning properly and passing.

I'll document this in an issue and dig a little deeper -- but it seems like a state format validation may be necessary. My hypothesis is that the incorrect state format was causing the read function to "error out" and return 0 records, resulting in a passing test.

lazebnyi reviewed (approved):
lgtm

alafanechere reviewed (commented):
@pnilan I re-ran the regression test tool on the PR (to benefit from the latest feature addition to the tool).
I noticed a change in the catalog and wanted to make sure you are aware of it and it's fine:
```
{
  "type_changes": {
    "root[1]['default_cursor_field']": {
      "old_type": "list",
      "new_type": "NoneType",
      "old_value": [
        "updated_at"
      ],
      "new_value": null
    },
    "root[1]['source_defined_cursor']": {
      "old_type": "bool",
      "new_type": "NoneType",
      "old_value": true,
      "new_value": null
    }
  },
  "iterable_item_removed": {
    "root[1]['supported_sync_modes'][1]": "incremental"
  }
}
```

pnilan commented:
@alafanechere Thanks for the catch. I believe this was a mistake on my end. I think it should be fixed now but would you be able to confirm?

alafanechere commented:
> @alafanechere Thanks for the catch. I believe this was a mistake on my end. I think it should be fixed now but would you be able to confirm?

I confirm the catalogs are matching now 👍

katmarkham commented on docs/integrations/sources/harvest.md:
```diff
@@ -88,6 +88,7 @@ The connector is restricted by the [Harvest rate limits](https://help.getharvest
 
 | Version | Date       | Pull Request                                             | Subject                                                                            |
 |:--------|:-----------|:---------------------------------------------------------|:-----------------------------------------------------------------------------------|
+| 1.0.0   | 2024-04-15 | [35863](https://github.com/airbytehq/airbyte/pull/35863) | Migrates connector to low code, Updates incremental substream state to per-partition state                                      |
```
low-code CDK

pnilan replied:
Updated
